File: bubus-ts/examples/simple.ts
#!/usr/bin/env -S node --import tsx
// Run: node --import tsx examples/simple.ts

import { BaseEvent, EventBus } from '../src/index.js'
import { z } from 'zod'

// 1) Define typed events with BaseEvent.extend(...)
const RegisterUserEvent = BaseEvent.extend('RegisterUserEvent', {
  email: z.string().email(),
  plan: z.enum(['free', 'pro']),
  // Handler return values for this event are validated against this schema.
  event_result_schema: z.object({
    user_id: z.string(),
    welcome_email_sent: z.boolean(),
  }),
})

const AuditEvent = BaseEvent.extend('AuditEvent', {
  message: z.string(),
})

async function main(): Promise<void> {
  const bus = new EventBus('SimpleExampleBus')

  // 2) Register a wildcard handler to observe every event flowing through this bus.
  bus.on('*', (event: BaseEvent) => {
    console.log(`[wildcard] ${event.event_type}#${event.event_id.slice(-8)}`)
  })

  // 3) Register by EventClass/factory (best type inference for payload + return type).
  bus.on(RegisterUserEvent, async (event) => {
    console.log(`[class handler] Creating account for ${event.email} (${event.plan})`)
    return {
      user_id: `user_${event.email.split('@')[0]}`,
      welcome_email_sent: true,
    }
  })

  // 4) Register by string event type (more dynamic, weaker compile-time checks).
  bus.on('AuditEvent', (event: InstanceType<typeof AuditEvent>) => {
    console.log(`[string handler] Audit log: ${event.message}`)
  })

  // 5) Intentionally return an invalid result shape.
  //    This compiles because string-based registration is best-effort, but will fail
  //    at runtime because RegisterUserEvent has event_result_schema enforcement.
  bus.on('RegisterUserEvent', () => {
    return { user_id: 123, welcome_email_sent: 'yes' } as unknown
  })

  // Dispatch a simple event handled by a string registration.
  await bus.emit(AuditEvent({ message: 'Starting simple bubus example' })).done()

  // Dispatch the typed event; one handler returns valid data, one returns invalid data.
  const register_event = bus.emit(
    RegisterUserEvent({
      email: 'ada@example.com',
      plan: 'pro',
    })
  )
  await register_event.done()

  // 6) Inspect per-handler results (completed vs error) from event.event_results.
  console.log('\nRegisterUserEvent handler outcomes:')
  for (const result of register_event.event_results.values()) {
    if (result.status === 'completed') {
      console.log(`- ${result.handler_name}: completed -> ${JSON.stringify(result.result)}`)
      continue
    }
    if (result.status === 'error') {
      const message = result.error instanceof Error ? result.error.message : String(result.error)
      console.log(`- ${result.handler_name}: error -> ${message}`)
      console.log(`  raw invalid return: ${JSON.stringify(result.raw_value)}`)
      continue
    }
    console.log(`- ${result.handler_name}: ${result.status}`)
  }

  // 7) Convenience getters for aggregate inspection.
  console.log('\nFirst valid parsed result:', register_event.first_result)
  console.log(`Total event errors: ${register_event.event_errors.length}`)
  for (const [index, error] of register_event.event_errors.entries()) {
    const message = error instanceof Error ? error.message : String(error)
    console.log(`  ${index + 1}. ${message}`)
  }

  await bus.waitUntilIdle()
  console.log('\n=== bus.logTree() ===')
  console.log(bus.logTree())
}

main().catch((error) => {
  console.error('Example failed:', error)
  process.exitCode = 1
})

File: examples/simple.py
#!/usr/bin/env -S uv run python
"""Run: uv run python examples/simple.py"""

import asyncio
from typing import Any, Literal

from pydantic import BaseModel

from bubus import BaseEvent, EventBus


class RegisterUserResult(BaseModel):
    user_id: str
    welcome_email_sent: bool


class RegisterUserEvent(BaseEvent[RegisterUserResult]):
    email: str
    plan: Literal['free', 'pro']
    event_result_type: Any = RegisterUserResult


class AuditEvent(BaseEvent[None]):
    message: str


def short_id(event_id: str) -> str:
    return event_id[-8:]


async def main() -> None:
    bus = EventBus(name='SimpleExampleBus')

    try:
        # 1) Observe every event via wildcard registration.
        def on_wildcard(event: BaseEvent[Any]) -> None:
            print(f'[wildcard] {event.event_type}#{short_id(event.event_id)}')

        bus.on('*', on_wildcard)

        # 2) Register a typed class handler.
        async def on_register_user(event: RegisterUserEvent) -> RegisterUserResult:
            print(f'[class handler] Creating account for {event.email} ({event.plan})')
            return RegisterUserResult(
                user_id=f"user_{event.email.split('@', maxsplit=1)[0]}",
                welcome_email_sent=True,
            )

        bus.on(RegisterUserEvent, on_register_user)

        # 3) Register by string event type.
        def on_audit(event: AuditEvent) -> None:
            print(f'[string handler] Audit log: {event.message}')

        bus.on('AuditEvent', on_audit)

        # 4) Intentionally return an invalid shape for runtime result validation.
        def on_register_user_invalid(_event: RegisterUserEvent) -> object:
            return {'user_id': 123, 'welcome_email_sent': 'yes'}

        bus.on('RegisterUserEvent', on_register_user_invalid)

        # Dispatch a simple event handled by string registration.
        await bus.emit(AuditEvent(message='Starting simple bubus example'))

        # Dispatch typed event; one handler is valid, one is invalid.
        register_event = bus.emit(
            RegisterUserEvent(
                email='ada@example.com',
                plan='pro',
            )
        )
        await register_event

        print('\nRegisterUserEvent handler outcomes:')
        for result in register_event.event_results.values():
            if result.status == 'completed':
                print(f'- {result.handler_name}: completed -> {result.result!r}')
                continue
            if result.status == 'error':
                message = str(result.error) if result.error is not None else 'unknown error'
                print(f'- {result.handler_name}: error -> {message}')
                continue
            print(f'- {result.handler_name}: {result.status}')

        first_valid = await register_event.event_result(raise_if_any=False, raise_if_none=False)
        all_errors = [result.error for result in register_event.event_results.values() if result.error is not None]

        print(f'\nFirst valid parsed result: {first_valid!r}')
        print(f'Total event errors: {len(all_errors)}')
        for index, error in enumerate(all_errors, start=1):
            print(f'  {index}. {error}')

        await bus.wait_until_idle()
        print('\n=== bus.log_tree() ===')
        print(bus.log_tree())
    finally:
        await bus.stop(clear=True, timeout=0)


if __name__ == '__main__':
    asyncio.run(main())
